### PR TITLE
Diff asset tracks cost changes

### PR DESCRIPTION
**Pull request by nickcurie**
## What does this PR change?
* Diff asset function will keep track of total cost changes

## Does this PR relate to any other PRs?
* [Yes](https://github.com/kubecost/docs/pull/301)

## How will this PR impact users?
* If an asset changed cost between two windows, it will be returned back to the user

## Does this PR address any GitHub or Zendesk issues?
* No

## How was this PR tested?
* Unit tests

## Does this PR require changes to documentation?
* Yes

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next Opencost release? If not, why not?
* No, don't have access

### CODE REVIEW

File: pkg/kubecost/asset.go [hunk 1]
```diff
@@ -2902,23 +2903,26 @@ type Diff[T any] struct {
...
 
 	for assetKey1, asset1 := range before.assets {
-		if _, ok := after.assets[assetKey1]; !ok {
+		if asset2, ok := after.assets[assetKey1]; !ok {
 			d := Diff[Asset]{asset1, DiffRemoved}
-			changedItems = append(changedItems, d)
+			changedItems[assetKey1] = d
+		} else if asset1.TotalCost() != asset2.TotalCost() {
+			d := Diff[Asset]{asset1, DiffChanged}
+			changedItems[assetKey1] = d
```
Comment (michaelmdresser) [inline pkg/kubecost/asset.go]: Since we have two assets to compare now, what do you think about the `Diff` type including a before and after for `T`, so users can see why the change happened? Like so:

```go
type Diff[T any] struct {
    Before T
    After T
    Kind DiffKind
}
```

We have this opportunity because we're messing with the API schema a bit anyway.

**Comment by sundeepghuman** [comment]
Should we have a threshold for cost change here?

File: pkg/kubecost/asset.go [hunk 2]
```diff
@@ -2902,27 +2904,34 @@ type Diff[T any] struct {
...
 
-// DiffAsset takes two AssetSets and returns a slice of Diffs by checking
-// the keys of each AssetSet. If a key is not found, a Diff is generated
-// and added to the slice.
-func DiffAsset(before, after *AssetSet) []Diff[Asset] {
-	changedItems := []Diff[Asset]{}
+// DiffAsset takes two AssetSets and returns a map of keys to Diffs by checking
+// the keys of each AssetSet. If a key is not found or is found with a different total cost,
+// a Diff is generated and added to the map.
+func DiffAsset(before, after *AssetSet, ratioCostChange float64) (map[string]Diff[Asset], error) {
```
Comment (michaelmdresser) [inline pkg/kubecost/asset.go]: Let's add a bit to the comment about the cost change. Something like "If an asset exists in before and after and the total cost after is greater than ratio*before total cost, the asset will be added to the diff"

**Review by michaelmdresser — COMMENTED** [review]
Thanks for adding the cost change parameter as suggested by Sundeep.

The functionality looks solid to me, but now that we're truly comparing 2 assets, I think adding the Before and After field on the Diff type as I suggested earlier is a good idea here. That way, users can inspect why a cost change occurred. What do you think?

File: pkg/kubecost/asset.go (hunk 1 above)
Comment (nickcurie) [reply pkg/kubecost/asset.go]: I agree that this is necessary, will work on getting this added

**Review by michaelmdresser — APPROVED** [review]

**Review by Sean-Holcomb — APPROVED** [review]
Looks good, nice unit tests
